src/collector/worker.rs

```diff
@@ -0,0 +1,142 @@
+// vectorchord-inquiry@tensorchord.ai
+//
+// Copyright (c) 2025 TensorChord Inc.
+use crate::collector::types::PgRefCell;
+use std::cell::RefMut;
+use std::fs;
+use std::path::Path;
+
+// Safety: The directory name must start with "pgsql_tmp" to be excluded by pg_basebackup
+const COLLECTOR_DIR: &str = "pgsql_tmp_vchord_sampling";
+const COLLECTOR_VERSION: u32 = 1;
+
+static CONNECTION: PgRefCell<Option<rusqlite::Connection>> =
+    PgRefCell::<Option<rusqlite::Connection>>::new(None);
+
+fn get<'a>() -> Option<RefMut<'a, rusqlite::Connection>> {
+    if unsafe { !pgrx::pg_sys::IsBackendPid(pgrx::pg_sys::MyProcPid) } {
+        return None;
+    }
+    let database_oid = unsafe { pgrx::pg_sys::MyDatabaseId.to_u32() };
+    if database_oid == 0 {
+        return None;
+    }
+    let mut connection = CONNECTION.borrow_mut();
+    if connection.is_none()
+        && let Err(err) = || -> rusqlite::Result<()> {
+            if !Path::new(COLLECTOR_DIR).exists() {
+                let _ = fs::create_dir_all(COLLECTOR_DIR);
+            }
+            let p = format!("{COLLECTOR_DIR}/database_{database_oid}.sqlite");
+            let mut conn = rusqlite::Connection::open(&p)?;
+            conn.pragma_update(Some("main"), "journal_mode", "WAL")?;
+            let tx = conn.transaction()?;
+            let version: u32 = tx
+                .pragma_query_value(Some("main"), "user_version", |row| row.get(0))
+                .unwrap_or(COLLECTOR_VERSION);
+            if version != COLLECTOR_VERSION && version != 0 {
+                let mut statement = tx.prepare(
+                    "SELECT name FROM sqlite_master WHERE type='table' AND name LIKE 'index_%';",
+                )?;
+                let tables = statement.query_map((), |row| row.get::<usize, String>(0))?;
+                for name in tables.into_iter().flatten() {
+                    let drop_statement = format!("DROP TABLE IF EXISTS {name}");
+                    tx.execute(&drop_statement, ())?;
+                }
+            }
+
+            tx.pragma_update(Some("main"), "user_version", COLLECTOR_VERSION)?;
+            tx.commit()?;
+            let _ = connection.insert(conn);
+            Ok(())
+        }()
+    {
+        pgrx::debug1!("Collector: Error initializing database: {}", err);
+        return None;
+    }
+    RefMut::filter_map(connection, |c| c.as_mut()).ok()
+}
+
+pub fn push(index: u32, sample: &str, max_records: u32) {
+    let mut connection = match get() {
+        Some(c) => c,
+        None => return,
+    };
+    let init_statement = format!(
+        "
+        CREATE TABLE IF NOT EXISTS index_{index} (sample TEXT, create_at REAL);
+        CREATE INDEX IF NOT EXISTS i ON index_{index} (create_at);
+        "
+    );
+    let insert_statement =
+        format!("INSERT INTO index_{index} (sample, create_at) VALUES (?1, unixepoch('subsec'))");
+    let count_statement = format!("SELECT COUNT(create_at) FROM index_{index}");
```

Comment:
This should not be performed. This will only result in what was originally a full table scan once now becoming a full table scan twice.